### PR TITLE
Linq tweaks

### PR DESCRIPTION
**Pull request by drewnoakes**
Various linq-related improvements

Changes proposed in this pull request:
 - Simplify `Where`/`First` style expressions
 - Simplify `SelectMany`/`Count` expression (potentially much fewer iterations required)
 - Don't use `Count()` if `Count`/`Length` available
 - Replace loops with linq expressions
 
What did I do to test the code and ensure quality:
 - Careful review
 - Manual testing
 - 

Has been tested on:
 - GIT 2.16
 - Windows 10

### CODE REVIEW

File: GitUI/UserControls/FileStatusList.cs [hunk 1]
```diff
@@ -488,7 +488,7 @@ public int UnfilteredItemsCount()
             }
             else
             {
-                return GitItemStatusesWithParents.SelectMany(pair => pair.Value).Count();
+                return GitItemStatusesWithParents.Sum(pair => pair.Value.Count);
```
Comment (RussKie) [inline GitUI/UserControls/FileStatusList.cs]: Are you sure the change is equivalent to the original?
`SelectMany` returns an `IEnumerable<whatever>` then does a count over the collection. The proposal sums up counts...
I'm looking at in a browser, and it is not exactly obvious.

File: GitCommands/ExternalLinks/ExternalLinkRevisionParser.cs [hunk 2]
```diff
@@ -99,18 +99,16 @@ private static IEnumerable<ExternalLink> ParseRevision(GitRevision revision, Ext
 
             if (definition.SearchInParts.Contains(ExternalLinkDefinition.RevisionPart.LocalBranches))
             {
-                foreach (var head in revision.Refs.Where(b => !b.IsRemote))
-                {
-                    links.Add(ParseRevisionPart(revision, definition, remoteMatch, head.LocalName));
-                }
+                links.AddRange(revision.Refs
+                    .Where(b => !b.IsRemote)
+                    .Select(head => ParseRevisionPart(revision, definition, remoteMatch, head.LocalName)));
```
Comment (RussKie) [inline GitCommands/ExternalLinks/ExternalLinkRevisionParser.cs]: I don't like this indentation - it gives impression `.Where` and `.Select` apply `links`.
Same applies to the change below.

File: GitCommands/Git/GitModule.cs [hunk 3]
```diff
@@ -3766,29 +3766,25 @@ public IList<GitItemStatus> GetCombinedDiffFileList(string shaOfMergeCommit)
         {
             var fileList = RunGitCmd("diff-tree --name-only -z --cc --no-commit-id " + shaOfMergeCommit);
 
-            var ret = new List<GitItemStatus>();
             if (string.IsNullOrWhiteSpace(fileList))
             {
-                return ret;
+                return new GitItemStatus[0];
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs]: Pure curiosity, is `Enumerable.Empty<GitStatus>()` worse? 
Looking at it in the reflector, it creates a singleton instance of `GitItemStatus[0]`....

File: GitUI/UserControls/FileStatusList.cs (hunk 1 above)
Comment (drewnoakes) [reply GitUI/UserControls/FileStatusList.cs]: I believe it's equivalent but faster. Let's look at an example.

```c#
var source = new int[][]
{
    Enumerable.Range(1000).ToArray(),
    Enumerable.Range(5000).ToArray()
};

// flattens to a sequence and enumerates 6000 items

source.SelectMany(a => a).Count() == 6000

// enumerates two items, sums 1000 and 5000

source.Sum(a => a.Length) == 6000
```

So yes, it's the same but less computationally expensive. Especially as `SelectMany` returns a collection for which the usual `Enumerable.Count` performance short-cuts (testing for `ICollection.Count`) do not apply due to its laziness.

The magnitude of gain in this change would be less than in my contrived example, as many items will have one parent. But there is still less iteration going on overall due to the reduction in number of operators.

File: GitCommands/Git/GitModule.cs (hunk 3 above)
Comment (drewnoakes) [reply GitCommands/Git/GitModule.cs]: No, this should have been changed to `Enumerable.Empty<GitItemStatus>()`. I'll push a fix shortly. Thanks.

And yes `Enumerable.Empty<>` returns a singleton empty array (as does `Array.Empty<>`) so assuming you use the empty array for a type more than once you can reduce allocations.

File: GitCommands/ExternalLinks/ExternalLinkRevisionParser.cs (hunk 2 above)
Comment (drewnoakes) [reply GitCommands/ExternalLinks/ExternalLinkRevisionParser.cs]: Thanks. I've pushed an improvement.

File: GitCommands/Git/GitModule.cs (hunk 3 above)
Comment (drewnoakes) [reply GitCommands/Git/GitModule.cs]: Fixed. Needed to be `Array.Empty<>` here as the return type is `IList<>`.

Once these `GitModule` PRs are finalised I'll go through and change a bunch of these `IList<T>` and `T[]` return types to `IReadOnlyList<>`. Maybe some of the `IEnumerable<T>` as well, in cases where the collection is already materialised.

**Pull request merged**